tests/unit/test_config_provider.py

```diff
@@ -0,0 +1,418 @@
+        )
+
+    def test_does_provide_value_if_variable_exists_in_list(self):
+        self.assert_does_provide(
+            names=['FOO'],
+            env={
+                'FOO': 'bar',
+            },
+            expected_value='bar',
+        )
+
+    def test_does_provide_first_non_none_value_first(self):
+        self.assert_does_provide(
+            names=['FOO', 'BAR'],
+            env={
+                'FOO': 'baz',
+            },
+            expected_value='baz',
+        )
+
+    def test_does_provide_first_non_none_value_second(self):
+        self.assert_does_provide(
+            names=['FOO', 'BAR'],
+            env={
+                'BAR': 'baz',
+            },
+            expected_value='baz',
+        )
+
+    def test_does_provide_none_if_all_list_variables_are_none(self):
+        self.assert_does_provide(
+            names=['FOO', 'BAR'],
+            env={},
+            expected_value=None,
+        )
+
+    def test_does_provide_first_value_when_both_exist(self):
+        self.assert_does_provide(
+            names=['FOO', 'BAR'],
+            env={
+                'FOO': 'baz',
+                'BAR': 'buz',
+            },
+            expected_value='baz',
+        )
+
+
+class TestConfigPropertyProvider(unittest.TestCase):
+    def assert_provides_value(self, config_file_values, config_var_name,
+                              expected_value):
+        fake_session = mock.Mock(spec=session.Session)
+        fake_session.get_scoped_config.return_value = config_file_values
+        property_provider = ConfigPropertyProvider(
+            config_var_name=config_var_name,
+            session=fake_session,
+        )
+        value = property_provider.provide()
+        self.assertEqual(value, expected_value)
+
+    def test_can_provide_value(self):
+        self.assert_provides_value(
+            config_file_values={
+                'foo': 'bar'
+            },
+            config_var_name='foo',
+            expected_value='bar',
+        )
+
+    def test_does_provide_none_if_var_not_in_config(self):
+        self.assert_provides_value(
+            config_file_values={
+                'foo': 'bar'
+            },
+            config_var_name='no_such_var',
+            expected_value=None,
+        )
+
+
+def _make_provider_that_returns(return_value):
+    provider = mock.Mock(spec=BaseProvider)
+    provider.provide.return_value = return_value
+    return provider
+
+
+def _make_providers_that_return(return_values):
+    mocks = []
+    for return_value in return_values:
+        provider = _make_provider_that_returns(return_value)
+        mocks.append(provider)
+    return mocks
+
+
+def assert_chain_does_provide(providers, expected_value):
+    provider = ChainProvider(
+        providers=providers,
+    )
+    value = provider.provide()
+    assert_equal(value, expected_value)
+
+
+def test_chain_provider():
+    # Each case is a tuple with the first element being the expected return
```

Comment:
Not too picky about this, but if you wanted you could copy a bit of what we do in [testing s3 addressing](https://github.com/boto/botocore/blob/7052b0f91fca61bf5206ec73240a99094fd2207a/tests/functional/test_s3.py#L855) where we have a class to specify cases. The nice part about this is that you can have parameter names so we would not have to explain the parameters in the comments and they would be self-documenting.

Reply:
Personally I like the way I did it better for this case since its super easy to see the chain and the result it should give with no extra verbosity.